util/timer.h

```diff
@@ -128,16 +141,22 @@ class Timer {
       }
 
       if (current_fn->next_run_time_us <= env_->NowMicros()) {
+        executing_task_ = true;
+        mutex_.Unlock();
         // Execute the work
         current_fn->fn();
+        mutex_.Lock();
+        executing_task_ = false;
+        cond_var_.SignalAll();
 
         // Remove the work from the heap once it is done executing.
         // Note that we are just removing the pointer from the heap. Its
         // memory is still managed in the map (as it holds a unique ptr).
         // So current_fn is still a valid ptr.
         heap_.pop();
 
-        if (current_fn->repeat_every_us > 0) {
+        // current_fn may be cancelled already.
+        if (current_fn->IsValid() && current_fn->repeat_every_us > 0) {
```

Comment:
should we add check for running, like:
```suggestion
        if (running_ && current_fn->IsValid() && current_fn->repeat_every_us > 0) {
```

Reply:
It's a good practice to check due to the preceding mutex release. In this case, if `current_fn->IsValid()` is true, then `running_` must still be true. This is because `running_` is set to `false` and all `fn->valid` are set to `false` in `Shutdown` before releasing the mutex.
I can add this.